### PR TITLE
fix: potential edge bugs of retry

### PR DESCRIPTION
**Pull request by CSWYF3634076**
#### What type of PR is this?
fix

#### Check the PR title.

- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.

#### (Optional) Translate the PR title into Chinese.
修复retry潜在的bug

#### (Optional) More detail description for this PR(en: English/zh: Chinese).

en:
1. In the case of the direct initialization definition &http1. HostClient {} in the picture, when the number of retries is greater than 1, the delay time will be calculated, but the retryConfig. DelayPolicy of retryConfig is nil, and when the retryConfig. DelayPolicy function is executed, a nil pointer error will be reported
2. When the MaxJitter of retryConfig is 0, a panic will occur when using retry.RandomDelayPolicy to generate random numbers
3. The negative number moves to the left will become very large, which is unreasonable

zh(optional):
1.图片中直接初始化定义&http1.HostClient{}的情况，retry次数大于1时，就会计算延迟时间，但retryConfig的retryConfig.DelayPolicy为nil，这时候执行retryConfig.DelayPolicy函数就会报空指针错误
2.在retryConfig的MaxJitter为0时，使用retry.RandomDelayPolicy生成随机数时会panic
3.负数左移会变的特别大，不合理
image

### CODE REVIEW

File: pkg/app/client/retry/retry.go [hunk 1]
```diff
@@ -63,11 +63,17 @@ func FixedDelayPolicy(_ uint, _ error, retryConfig *Config) time.Duration {
 
 // RandomDelayPolicy is a DelayPolicyFunc which picks a random delay up toRetryConfig.MaxJitter
 func RandomDelayPolicy(_ uint, _ error, retryConfig *Config) time.Duration {
+	if retryConfig.MaxJitter <= 0 {
```
Comment (li-jin-gou) [inline pkg/app/client/retry/retry.go:66]: math/rand => https://github.com/bytedance/gopkg/tree/develop/lang/fastrand

File: pkg/app/client/retry/retry.go [hunk 2]
```diff
@@ -63,11 +63,17 @@ func FixedDelayPolicy(_ uint, _ error, retryConfig *Config) time.Duration {
...
+		return 0 * time.Millisecond
+	}
 	return time.Duration(rand.Int63n(int64(retryConfig.MaxJitter)))
 }
 
 // BackOffDelayPolicy is a DelayPolicyFunc which exponentially increases delay between consecutive retries
 func BackOffDelayPolicy(attempts uint, _ error, retryConfig *Config) time.Duration {
+	if retryConfig.Delay <= 0 {
```
Comment (welkeyever) [inline pkg/app/client/retry/retry.go:74]: 这个新加的判断逻辑在注释里也描述一下哈，`Delay `、`RandomDelayPolicy`同

File: pkg/app/client/retry/retry.go (hunk 1 above)
Comment (CSWYF3634076) [reply pkg/app/client/retry/retry.go:67]: done

File: pkg/app/client/retry/retry.go (hunk 2 above)
Comment (CSWYF3634076) [reply pkg/app/client/retry/retry.go:76]: done

Comment (welkeyever) [reply pkg/app/client/retry/retry.go:77]: em...是加载函数注释上面哈：因为这个行为和目前的描述其实是有diff的（属于一个特异场景），需要用户在使用函数的时候感知（写在内部的这个注释除非直接看源码，正常的ide都不会显示的哈）；核心目的是让用户在使用这个函数的清时候楚这个行为。

Comment (CSWYF3634076) [reply pkg/app/client/retry/retry.go:75]: done

File: pkg/app/client/retry/retry.go [hunk 3]
```diff
@@ -61,13 +62,19 @@ func FixedDelayPolicy(_ uint, _ error, retryConfig *Config) time.Duration {
...
 func RandomDelayPolicy(_ uint, _ error, retryConfig *Config) time.Duration {
-	return time.Duration(rand.Int63n(int64(retryConfig.MaxJitter)))
+	if retryConfig.MaxJitter <= 0 {
+		return 0 * time.Millisecond
+	}
+	return time.Duration(fastrand.Int63n(int64(retryConfig.MaxJitter)))
 }
 
-// BackOffDelayPolicy is a DelayPolicyFunc which exponentially increases delay between consecutive retries
+// BackOffDelayPolicy is a DelayPolicyFunc which exponentially increases delay between consecutive retries,if the InitDelay <= 0 then return 0 , prevent negative number left shift
```
Comment (welkeyever) [inline pkg/app/client/retry/retry.go:73]: `if`前面缺了个空格，`InitDelay `和入参没有对应上，应该是retryConfig.InitDelay? `prevent negative number left shift`这个描述涉及到的是内部实现问题（业务部感知），可以直接去掉；
整体上可以描述为：
`If the xxx is less than or equal to 0, the final delay is 0`
其他的可以按照这个逻辑更新一下哈

Comment (CSWYF3634076) [reply pkg/app/client/retry/retry.go]: done

File: pkg/app/client/retry/retry.go [hunk 4]
```diff
@@ -94,8 +101,12 @@ func CombineDelay(delays ...DelayPolicyFunc) DelayPolicyFunc {
 	}
 }
 
-// Delay generate the delay time required for the current retry
+// Delay generate the delay time required for the current retry config , if the retryConfig.DelayPolicy == nil , the final delay is 0
```
Comment (welkeyever) [inline pkg/app/client/retry/retry.go:104]: `config`后面好像多了个空格，逗号前面不需要空格哈

Comment (welkeyever) [reply pkg/app/client/retry/retry.go:104]: 其他的也看看有个上有个0后面应该也是多了个空格

Comment (CSWYF3634076) [reply pkg/app/client/retry/retry.go]: done

**Review by welkeyever — APPROVED** [review]
lgtm

**Review by li-jin-gou — APPROVED** [review]

**Pull request merged**